File: src/main/java/no/novari/flyt/egrunnerverv/gateway/slack/SlackAlertService.java
package no.novari.flyt.egrunnerverv.gateway.slack;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.HashMap;
import java.util.Map;

@Service
@Slf4j
public class SlackAlertService {

    @Value("${fint.org-id}")
    private String orgId;

    @Value("${fint.application-id}")
    private String applicationId;

    @Value("${slack.webhook.url}")
    private String slackWebhookUrl;

    private final WebClient webClient;

    public SlackAlertService(
            @Qualifier("slackWebClient") WebClient webClient
    ) {
        this.webClient = webClient;
    }

    public void sendMessage(String message) {
        Map<String, String> payload = new HashMap<>();
        String formattedMessage = formatMessageWithPrefix(message);
        payload.put("text", formattedMessage);

        webClient.post()
                .uri(slackWebhookUrl)
                .bodyValue(payload)
                .retrieve()
                .bodyToMono(Void.class)
                .block();
    }

    private String formatMessageWithPrefix(String message) {
        return orgId + "-" + applicationId + "-" + message;
    }
}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/slack/SlackAlertService.kt
package no.novari.flyt.egrunnerverv.gateway.slack

import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClient
import org.springframework.web.client.RestClientResponseException

@Service
class SlackAlertService(
    @param:Value("\${fint.org-id}") private val orgId: String,
    @param:Value("\${fint.application-id}") private val applicationId: String,
    @param:Value("\${slack.webhook.url}") private val slackWebhookUrl: String,
    @param:Qualifier("slackRestClient") private val restClient: RestClient,
) {
    fun sendMessage(message: String) {
        val payload = mapOf("text" to formatMessageWithPrefix(message))
        try {
            restClient
                .post()
                .uri(slackWebhookUrl)
                .body(payload)
                .retrieve()
                .toBodilessEntity()
        } catch (ex: RestClientResponseException) {
            logger.warn("Slack webhook failed with status={}", ex.statusCode, ex)
        } catch (ex: RuntimeException) {
            logger.warn("Slack webhook failed", ex)
        }
    }

    private fun formatMessageWithPrefix(message: String): String = "$orgId-$applicationId-$message"

    companion object {
        private val logger = org.slf4j.LoggerFactory.getLogger(SlackAlertService::class.java)
    }
}
